hazelcast-sql/src/test/java/com/hazelcast/sql/impl/expression/predicate/CaseExpressionTest.java

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright (c) 2008-2020, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.sql.impl.expression.predicate;
+
+import com.hazelcast.sql.SqlService;
+import com.hazelcast.sql.impl.SqlDataSerializerHook;
+import com.hazelcast.sql.impl.expression.ConstantExpression;
+import com.hazelcast.sql.impl.expression.Expression;
+import com.hazelcast.sql.impl.expression.ExpressionTestBase;
+import com.hazelcast.sql.impl.expression.SimpleExpressionEvalContext;
+import com.hazelcast.sql.impl.type.QueryDataType;
+import com.hazelcast.test.HazelcastParallelClassRunner;
+import com.hazelcast.test.annotation.ParallelJVMTest;
+import com.hazelcast.test.annotation.QuickTest;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+
+import static com.hazelcast.sql.impl.type.QueryDataType.BIGINT;
+import static com.hazelcast.sql.impl.type.QueryDataType.BOOLEAN;
+import static com.hazelcast.sql.impl.type.QueryDataType.INT;
+import static org.junit.Assert.assertEquals;
+
+@RunWith(HazelcastParallelClassRunner.class)
+@Category({QuickTest.class, ParallelJVMTest.class})
+public class CaseExpressionTest extends ExpressionTestBase {
+
+    @Test
+    public void testEndToEnd() {
+        SqlService sql = createEndToEndRecords();
+        assertRows(query(sql, "select __key from records where case boolean1 when true then false when false then true end"),
+                keyRange(0, 500));
+        assertRows(query(sql, "select __key from records where case not boolean1 when true then false when false then true end"),
+                keyRange(500, 1000));
+        assertRows(query(sql,
+                "select __key, case when __key < 500 then double1 when __key < 1000 then int1 else decimal1 end from records"),
+                keyRange(0, 1000, 5000, 6000), k -> {
+                    if (k < 500) {
+                        return 2000.1 + k;
+                    } else if (k < 1000) {
+                        return 1000.0 + k;
+                    } else {
+                        return k == 5000 ? 9001.0 : null;
+                    }
+                });
+        assertQueryThrows(sql, "select * from records where case boolean1 when true then string1 end",
+                "where clause must be a condition");
```

Comment:
This error message is not precise, better would be "must be a boolean expression". You can use a boolean literal, but is that a condition?

However, it comes from calcite code, so if it's not easy to replace it, maybe it's not worth the trouble:
```
	at org.apache.calcite.sql.validate.SqlValidatorImpl.validateWhereOrOn(SqlValidatorImpl.java:4200)
	at org.apache.calcite.sql.validate.SqlValidatorImpl.validateWhereClause(SqlValidatorImpl.java:4184)
	at org.apache.calcite.sql.validate.SqlValidatorImpl.validateSelect(SqlValidatorImpl.java:3513)
	at org.apache.calcite.sql.validate.SelectNamespace.validateImpl(SelectNamespace.java:60)
	at org.apache.calcite.sql.validate.AbstractNamespace.validate(AbstractNamespace.java:84)
	at org.apache.calcite.sql.validate.SqlValidatorImpl.validateNamespace(SqlValidatorImpl.java:1110)
	at org.apache.calcite.sql.validate.SqlValidatorImpl.validateQuery(SqlValidatorImpl.java:1084)
	at org.apache.calcite.sql.SqlSelect.validate(SqlSelect.java:232)
	at org.apache.calcite.sql.validate.SqlValidatorImpl.validateScopedExpression(SqlValidatorImpl.java:1059)
	at org.apache.calcite.sql.validate.SqlValidatorImpl.validate(SqlValidatorImpl.java:766)
	at com.hazelcast.sql.impl.calcite.parse.QueryParser.parse(QueryParser.java:55)
```

Reply:
case removed in c1a5855c8212e429b5d7245a1906713af01e624b and 7f29e69e24bcd44dec551b9d255171a48bdda0ec, https://github.com/hazelcast/hazelcast/issues/17329 created

Reply:
This is not specific to CASE expression.